src/Cedar/Session.c

```diff
@@ -1949,23 +1956,63 @@ SESSION *NewClientSessionEx(CEDAR *cedar, CLIENT_OPTION *option, CLIENT_AUTH *au
 	{
 		s->ClientAuth->ClientX = CloneX(s->ClientAuth->ClientX);
 	}
-  if (s->ClientAuth->ClientK != NULL)
-  {
-    if (s->ClientAuth->AuthType != CLIENT_AUTHTYPE_OPENSSLENGINE)
-    {
-      s->ClientAuth->ClientK = CloneK(s->ClientAuth->ClientK);
-    }
-    else
-    {
-      s->ClientAuth->ClientK = OpensslEngineToK(s->ClientAuth->OpensslEnginePrivateKeyName, s->ClientAuth->OpensslEngineName);
-    }
-  }
+	if (s->ClientAuth->ClientK != NULL)
+	{
+		if (s->ClientAuth->AuthType != CLIENT_AUTHTYPE_OPENSSLENGINE)
+		{
+			s->ClientAuth->ClientK = CloneK(s->ClientAuth->ClientK);
+		}
+		else
+		{
+			s->ClientAuth->ClientK = OpensslEngineToK(s->ClientAuth->OpensslEnginePrivateKeyName, s->ClientAuth->OpensslEngineName);
+		}
+	}
 
 	if (StrCmpi(s->ClientOption->DeviceName, LINK_DEVICE_NAME) == 0)
 	{
 		// Link client mode
 		s->LinkModeClient = true;
 		s->Link = (LINK *)s->PacketAdapter->Param;
+		if (s->Link != NULL && s->Link->CheckServerCert && s->Link->Hub->HubDb != NULL)
+		{
+			// Enable SSL peer verification
+			s->SslOption = ZeroMalloc(sizeof(SSL_VERIFY_OPTION));
+			s->SslOption->VerifyPeer = true;
+			s->SslOption->AddDefaultCA = s->Link->AddDefaultCA;
+			s->SslOption->VerifyHostname = true;
+			s->SslOption->SavedCert = CloneX(s->Link->ServerCert);
+
+			// Copy trusted CA
+			LIST *o = s->Link->Hub->HubDb->RootCertList;
+			LockList(o);
+			{
+				s->SslOption->CaList = CloneXList(o);
+			}
+			UnlockList(o);

```

Comment:
```suggestion
			s->SslOption->CaList = CloneXList(o);

```
`CloneXList` already locks the list but why duplicate locking did not result in a deadlock? Is it because they are from the same thread?